### ISSUE TITLE
IsObjectHashed SCC entries

### ISSUE THREAD

**Issue by tajila** [eclipse-openj9/openj9#13029]
Currently, when `System::identityHashCode()` is called on an object we may need to grow a slot to preserve the identity hash value. This introduces a throughput cost as we need to reallocate the instance in larger memory.

To address this we can record the types of objects that are called with `System::identityHashCode()` in the SCC. Then on subsequent runs we can allocate a slot for those objects upfront so that we do not need to perform the reallocation at a later time. A JVM run may add new entries to the SCC if one doesn't already exist.

Phase 1: The simple approach
- As soon as `identityHashCode()` is called on an instance, we record the type of the instance and place an entry in the SCC (ie. requiresPreHashAlloc). 
- When loading classes we will tag ramClasses that have an entry in requiresPreHashAlloc table with J9ClassRequiresPreHashAlloc
- On allocations instances of these types will be allocated with `J9_GC_ALLOCATE_OBJECT_HASHED`
- JIT cthelper should be added to query if class type has J9ClassRequiresPreHashAlloc

Phase 2: Heuristic Approach
- Set a threshold for the number of times before an entry is made in the requiresPreHashAlloc table
- Monitor hashing metrics, if hashing is below a certain threshold invalidate entry in requiresPreHashAlloc table

In the short term we are concerned with Phase 1). This optimization should be toggle via a vmOption

**Comment by tajila** [comment]
Phase 1 needs to be done for 0.29

**Comment by tajila** [comment]
@dmitripivkine can we pre hash with the inline allocators? or do we need to use the out-of-line (mgcalloc) allocators?

**Comment by tajila** [comment]
FYI @hangshao0 @gacholio

**Comment by gacholio** [comment]
Prehash is not supported by the inline allocators at the moment (prehash is a request per out-of-line allocation, using the `J9_GC_ALLOCATE_OBJECT_HASHED` flag, not per class at the moment).

**Comment by gacholio** [comment]
Might be easiest to fail over to the OOL allocate in the interpreter if the class flag is set, and let the JIT do the inline prehash work.

I don't think we have the bits to have a "hash slot allocated but not filled" state, meaning that we will have to hash all of these objects on allocation, which is probably a significant perf penalty). The JIT could compile-time eliminate the test for positive hash codes.

**Comment by gacholio** [comment]
```
#define OBJECT_HEADER_HAS_BEEN_HASHED  0x20000
#define OBJECT_HEADER_HAS_BEEN_MOVED  0x10000
#define OBJECT_HEADER_HAS_BEEN_HASHED_MASK  0x30000
```

Given two bits, perhaps we can add the intermediate state - the states are:

- hash slot unallocated, not hashed, not moved (normal unhashed object)
- hash slot allocated, not hashed, not moved (new state for this feature)
- hash slot unallocated, hashed, not moved (normal object hashed)
- hash slot allocated, hashed, moved (same as hashed and moved today)

**Comment by hangshao0** [comment]
For SCC, a new type is needed in this list:
https://github.com/eclipse-openj9/openj9/blob/629fc4f8eb18604d91335f8fb73e323d953f8eee/runtime/oti/shcflags.h#L45-L58

`J9SHR_DATA_TYPE_CACHELET` is obsolete. We can probably rename it and reuse its value.

> On cold runs (runs that create SCC) as soon as identityHashCode() is called on an instance, we record the type of the instance and place an entry in the SCC (ie. requiresPreHashAlloc).

Are you saying that only the first JVM is able add such records to the SCC ? The SCC is enable by default in OpenJ9. For most users, the default SCC might be created by `Java -version`.
I guess we need to allow subsequent JVMs to add such info as well (if the record does not already exist in SCC).

**Comment by tajila** [comment]
> I guess we need to allow subsequent JVMs to add such record as well (if the record does not already exist in SCC).

I think what you are suggesting makes more sense, subsequent VMs should be able to add more entries

**Comment by pshipton** [comment]
I wonder if it should be tied to the GC startup hints, so there are different lists of hashed classes for different sets of command line options.

**Comment by fjeremic** [comment]
@0xdaryl would we always know at compile time the type for inline allocation? If so then we can also query the cthelper at compile time for `J9ClassRequiresPreHashAlloc` and we could potentially insert cross codegen IL to mark the object with `J9_GC_ALLOCATE_OBJECT_HASHED`. This way we could potentially avoid cross-codegen work for this.

**Comment by DanHeidinga** [comment]
> Currently, when `System::identityHashCode()` is called on an object we may need to grow a slot to preserve the identity hash value. This introduces a throughput cost as we need to reallocate the instance in larger memory.
 
This is a per-instance growth that only grows the object if it survives long enough to move.  For a tenured object, which moves rarely, we may go a long time before growing a slot.  For nursery objects, they may become garbage before we ever need to grow a slot.

Can you highlight some examples of the throughput cost?

> To address this we can record the types of objects that are called with `System::identityHashCode()` in the SCC. Then on subsequent runs we can allocate a slot for those objects upfront so that we do not need to perform the reallocation at a later time. A JVM run may add new entries to the SCC if one doesn't already exist.

This is a pretty big hammer to hit allocation with as it turns a per-instance property into a per-class property that affects all instances.  

There's a lot of enthusiasm for this change in the discussion so far but no data supplied to validate the throughput cost or speak to the `Phase 1 needs to be done for 0.29` deadline

**Comment by gacholio** [comment]
@fjeremic The class flag may be set during the run, so you would need to check every class that does not have the bit set, and patch the check out once it does.

**Comment by gacholio** [comment]
@DanHeidinga I'd also like to know the motivation for this, as it's looking more expensive all the time.

**Comment by dmitripivkine** [comment]
> ```
> #define OBJECT_HEADER_HAS_BEEN_HASHED  0x20000
> #define OBJECT_HEADER_HAS_BEEN_MOVED  0x10000
> #define OBJECT_HEADER_HAS_BEEN_HASHED_MASK  0x30000
> ```
> 
> 
>      
>   
> 
> Given two bits, perhaps we can add the intermediate state - the states are:
> 
> 
> 
> * hash slot unallocated, not hashed, not moved (normal unhashed object)
> 
> * hash slot allocated, not hashed, not moved (new state for this feature)
> 
> * hash slot unallocated, hashed, not moved (normal object hashed)
> 
> * hash slot allocated, hashed, moved (same as hashed and moved today)

FYI GC is used all four states of these two bits already to handle Scavenger Backout, some cases of CS etc. An extra state is used to determine case "object has been grown **this** Copy cycle"

**Comment by tajila** [comment]
> Can you highlight some examples of the throughput cost?

@0xdaryl @vijaysun-omr can you comment on this?

**Comment by vijaysun-omr** [comment]
@0xdaryl had documented the analysis around the user scenario motivating this discussion at : https://github.ibm.com/runtimes/openj9-jit-xopt/issues/439#issuecomment-29368090

**Comment by DanHeidinga** [comment]
> @0xdaryl had documented the analysis around the user scenario motivating this discussion at : https://github.ibm.com/runtimes/openj9-jit-xopt/issues/439#issuecomment-29368090

@vijaysun-omr that's not publicly visible.  Can you share the information here?  We need more detail in the open explaining the need for this change

**Comment by DanHeidinga** [comment]
Thanks for the context @vijaysun-omr.

> I wonder if it should be tied to the GC startup hints, so there are different lists of hashed classes for different sets of command line options.

@pshipton raises a good point.  The value of this is really application specific and should be tied to set of command line options as it may not apply to all uses.

Given the issue, I'd suggest an alternative approach that avoids using the SCC to start by adding a command line option to control this behaviour.  The option can be similar to the `-Xlockword:noLockword=java/lang/String,noLockword=....` option that provides a list of classes to apply it to.  New option could be something like `-XensureHasBeenHashed="foo/klass,bar/klass,...`

I'm also wondering if it would be better to handle this in the GC when tenuring rather than in the JIT/VM when allocating.  Could we set a bit in the `J9Class` - `AccClassGCSpecial` maybe? - that tells the GC to allocate and hash when tenuring the object?  It seems like it would be cheaper to have the GC handle this during tenure as that's when we would be growing the slot anyway.

@dmitripivkine does that seem reasonable?

**Comment by pshipton** [comment]
Note the new option should be additive, and not right most wins, so user's can add additional classes without affecting the ones in the options.default.

**Comment by gacholio** [comment]
So is the belief here that lots of instances are being hashed only after they've been tenured? The class flag solution seems like the best solution to avoid codegen changes that would likely result in a perf loss across the board.

The tenure change to check for object hashed or class flag "tenured instances must be hashed" seems pretty simple.

**Comment by dmitripivkine** [comment]
We are not in favour to add extra checks per copied object - this is most hot code in copy collector and it might be very expensive. In this case however we can try - we are reading class already so should not be expensive to check extra flag in the class (if it is loaded to cache line already).
I am worried slightly about effectiveness of this feature due timing nature of the problem _"objects have not been hashed by the time they are tenured"_. How it will behave for Balanced for example? Even in Gencon it might be different behaviours for various heap sizes. An object can be grown earlier, at the first copy (more generic but more expensive case obviously).
Should GC behave as "has been hashed" bit is set, so grow by hash slot if necessary and populate it?
BTW if we choose "grow at first copy" strategy there is enough to set "has been hashed" bit at object allocation time and no further modifications required

**Comment by gacholio** [comment]
> BTW if we choose "grow at first copy" strategy there is enough to set "has been hashed" bit at object allocation time and no further modifications required

Right - I had forgotten that we do not have to compute the hash in order to set that bit. This is by far the better solution given that the class flags will be tagged before the JIT can compile any code related to it, so there will be no need for a runtime check in the compiled code when the class being allocated is known (i.e. basically all the time).

We should update the out-of-line allocators in the GC and JIT/VM helpers to handle this case so only fully inlined allocations (JIT and interpreter) set the hashed bit.

**Comment by dmitripivkine** [comment]
> We should update the out-of-line allocators in the GC and JIT/VM helpers to handle this case so only fully inlined allocations (JIT and interpreter) set the hashed bit.

We can do it but is it a significant difference from the case just set bit and postpone hash code calculation to first copy collection - just asking?

**Comment by dmitripivkine** [comment]
ok, after thinking a little bit I agree with GAC

**Comment by gacholio** [comment]
Sorry, I was extremely unclear when I said "only fully inlined allocations set the hashed bit" - what I meant is that in the case where we fail over to the OOL allocation, the caller need not set the hash bit since the allocator will have done it.

Fully inline allocations will also need to set the bit (which will be a runtime check in the interpreter inline allocator).

**Comment by DanHeidinga** [comment]
If the behaviour is driven by a commandline option, then it will be statically known up front and Gac's approach sounds right.  The only question I still see is how AOT should behave.  Would the new command line option prevent AOT code from being shared between processes that had different option settings?  This is probably the easiest behaviour.  Alternatively, the AOT code can patch the bitset in if an allocation required it.

fyi @dsouzai for comment on the AOT aspects

**Comment by gacholio** [comment]
What does AOT do if the lockword configuration changes?

**Comment by dsouzai** [comment]
> Would the new command line option prevent AOT code from being shared between processes that had different option settings?

By default, no; we have to explicitly enforce consistency with the option.

> What does AOT do if the lockword configuration changes?

We verify that the hash of the lockword is the same:
https://github.com/eclipse-openj9/openj9/blob/53265fd814d22970bc01ca814eb1d7b913ea8027/runtime/compiler/runtime/RelocationRuntime.cpp#L1095-L1099

https://github.com/eclipse-openj9/openj9/blob/53265fd814d22970bc01ca814eb1d7b913ea8027/runtime/compiler/runtime/RelocationRuntime.cpp#L971-L985

If the idea is to have an option similar to the -Xlockword option then for AOT we'd just have to add a similar validation.

**Comment by 0xdaryl** [comment]
We had an offline discussion on this issue yesterday and I'll summarize the outcome of that discussion here.

TL;DR : we'll implement this feature with a command-line option solution

*   *   *   *   *

The original intent of this feature was to remember classes whose identity hashcode was taken in the SCC and then use that information in subsequent JVM invocations to set the "has-been-hashed" (HBH) bit when instances are allocated.  Expansion of the object to include the hashcode slot would be handled naturally using the existing GC mechanisms.  I think there was some earlier confusion in this issue around the scope of the feature proposed that eventually got sorted out through the discussion.

There have been a couple of different approaches proposed for setting the HBH bit, each with their own pros and cons.

The HBH bit could be set under direction of a command-line configuration option.  This is similar to the `-Xlockword` option already in use to proactively insert an extra field in an object instance for the lock word.  The advantage of this approach is that it limits the scope of the optimization to a user-prescribed set of classes and thereby reduces the overall risk on unseen workloads (e.g., footprint bloat of adding extra slots, or the overhead of computing and storing identity hashcodes during GC time).  However, the potential reward for this optimization is muted as the opportunity must be discovered and the affected classes added statically.

Alternatively, the potential classes could be discovered dynamically with hints left in the SCC for future JVM invocations.  This has the potential of identifying more classes where this would be beneficial based on objects whose identity hashcode was taken, and would be more adaptive to the user's application classes.  However, some heuristic-driven throttling may be required in order to focus the optimization on only the classes that would truly benefit.

A concern was raised that this seems like a problem specific to one specific customer's application and not seen generally in other applications (that we know about).  If this is a corner case then that suggests that a static, prescriptive solution might be the safest approach that minimizes the risk for other consumers.  We all agreed that the best way to move forward with this feature is to introduce it under control of a command-line option, and revisit a more dynamic approach should a second use case come about later.

We also discussed conducting some experiments to guide selecting an initial set of classes to specify statically as pre-hashed and to better understand this problem on workloads that OpenJ9 cares about.  Some instrumentation will be added to the identity hashcode computation path and the GC (the permanency of such instrumentation is TBD) to collect statistics on the classes involved and the location (nursery or tenure) of the objects being hashed.

The above discussion in this issue also raised a concern about the impact on AOT and whether implementing such a feature could cause inconsistencies between shared classes caches (and thereby limit their sharing).  Because the proposal here is to simply set a dynamic flag on the object when it is allocated that indicates that it has been hashed I don't believe there will be an impact to AOT.  The GC will be responsible for growing objects based on their dynamic state as it is now.  As the HBH bit does not indicate that the shape of the object is different I believe things will just proceed normally.

@tajila @vijaysun-omr @amicic @dmitripivkine @fjeremic

**Comment by dsouzai** [comment]
> Because the proposal here is to simply set a dynamic flag on the object when it is allocated that indicates that it has been hashed I don't believe there will be an impact to AOT.

This is true, but only if we always check the bit for all objects when generating AOT code. If the option specifies only `java/lang/String` in run 1, then if the compiler does not generate the check when it can prove an object is not a `java/lang/String`, then in run 2, if the option changes and includes more classes, the AOT code will be incorrect, because it will lack the code to check the bit for objects that are not `java/lang/String`. Even more fundamentally, if the option is not set in run 1 and the compiler does not generate any checks, then if the option is set in run 2, the AOT code can't be loaded.

So if we don't want the compiler to generate the code to check the bit for all objects (in an AOT compile) then we need a validation similar to Xlockword.

**Comment by hangshao0** [comment]
FYI @EricYangIBM

**Comment by EricYangIBM** [comment]
Should the ramClass flag be set in createramclass.cpp when the class is loaded? Also should subclasses of the specified classes inherit the flag?

**Comment by hangshao0** [comment]
> Should the ramClass flag be set in createramclass.cpp when the class is loaded?

Looks like it is the correct place.

> Also should subclasses of the specified classes inherit the flag?

Yes.

**Comment by hangshao0** [comment]
Adding comp:doc label as a new command line option is being added.

**Comment by pshipton** [comment]
Reopening since https://github.com/eclipse-openj9/openj9/pull/13338 was temporarily reverted.
